util/users.py

```diff
@@ -55,3 +56,34 @@ async def delete_user(identifier: user_models.UserIdentifier) -> None:
     if response.deleted_count == 0:
         detail = "User not found and could not be deleted"
         raise exceptions.UserNotFoundException(detail=detail)
+
+
+async def attempt_user_login(
+        login_form: user_models.UserLoginForm) -> user_models.UserLoginResponse:
+    """
+    Validates user login attempt based off
+    identifier and password. Will raise
+    404 UserNotFoundException if user does
+    not exist, or 422 InvalidPasswordException
+    if the user does exist but password is invalid.
+    """
+    user = await get_user_info_by_identifier(login_form.identifier)
+    password_matches = await check_user_password_matches(login_form, user)
+    if password_matches:
+        auth_token = await get_auth_token_from_user_data(user)
+        login_response = user_models.UserLoginResponse(jwt=auth_token)
+        return login_response
+    raise exceptions.InvalidPasswordException
+
+
+# fixme: change this to use hashed passwords once available
+async def check_user_password_matches(login_form: user_models.UserLoginForm,
+                                      user: user_models.User) -> bool:
+    return login_form.password == user.password
```

Comment:
Code to fix this is in upstream.

Also, since you offloaded the password comparison to this function, the refactor will be extremely easy, which is a great sign for the code quality.

Reply:
-I vaguely remember us discussing this, but I believe my code is up to date with upstream (I see last change was 10 days ago, Christian's push). Regardless, how can I pull master into my codebase without compromising what I've written? Was it as simple as doing a `git pull origin master`, or was it something else?

-`user.password` is going to be hashed, right? So the code would just be return user.check_password(login_form.password)? If that assumption is correct, that seems _almost_ overly complex, since user already has a `check_password` function. What do you think?

Reply:
Yeah your code is up to date with master (we ran `git pull origin master` last meeting), so this can implement the actual function.

Assumption is definitely correct, we never touch the hashed passwords directly so we would just be calling `user.check_password(login_form.password)` and returning that.

While plugging that call into the main `attempt_user_login` function is totally valid, it makes us call a method for a class that we don't have to interact with directly in the function, essentially making a mixed level of abstraction in the code. 

What if, for example, once we change the `User` class to also allow admin type users, we have a more complex password checking system that also checks for roles or something like that. 

By abstracting the actual call to the `User.check_password` method in it's own function, we would be able to rewrite that single function and never have to change the actual login function code.

Sometimes writing refactorable code seems a little extra and almost redundant, but adding clarity and flexibility is always worth the extra code. Writing more code is bad if it's just more code, but it almost never is.

Reply:
True, your example makes a lot of sense. I know it was a clarity/refactorability tradeoff, but considering the admin stuff is coming soon I think it makes a lot of sense to add this extra layer here.